File: packages/client-js/src/mcp-observer.ts
/**
 * MCP tool call observer — auto-instruments MCP client and server SDKs
 * to capture tool invocations, arguments, responses, latency, and errors.
 *
 * Writes to .trickle/mcp.jsonl as:
 *   { "kind": "mcp_tool_call", "tool": "fetch", "direction": "outgoing",
 *     "durationMs": 234.5, "args": {...}, "result": "...", ... }
 *
 * Zero code changes needed — intercepted via Module._load hook.
 */

import * as fs from 'fs';
import * as path from 'path';

let mcpFile: string | null = null;
let eventCount = 0;
const MAX_MCP_EVENTS = 1000;
const TRUNCATE_LEN = 500;

function getMcpFile(): string {
  if (mcpFile) return mcpFile;
  const dir = process.env.TRICKLE_LOCAL_DIR || path.join(process.cwd(), '.trickle');
  try { fs.mkdirSync(dir, { recursive: true }); } catch {}
  mcpFile = path.join(dir, 'mcp.jsonl');
  return mcpFile;
}

interface McpEvent {
  kind: 'mcp_tool_call';
  tool: string;
  direction: 'outgoing' | 'incoming';
  durationMs: number;
  args: unknown;
  resultPreview: string;
  isError: boolean;
  errorMessage?: string;
  timestamp: number;
}

function writeMcpEvent(event: McpEvent): void {
  if (eventCount >= MAX_MCP_EVENTS) return;
  eventCount++;
  try {
    fs.appendFileSync(getMcpFile(), JSON.stringify(event) + '\n');
  } catch {}
}

function truncate(s: string, len = TRUNCATE_LEN): string {
  if (!s) return '';
  return s.length > len ? s.substring(0, len) + '...' : s;
}

function sanitizeArgs(args: unknown): unknown {
  if (args === null || args === undefined) return null;
  try {
    const s = JSON.stringify(args);
    return s.length > 1000 ? JSON.parse(s.substring(0, 1000) + '"}') : args;
  } catch { return String(args).substring(0, 200); }
}

function extractResultPreview(result: any): string {
  if (!result) return '';
  // MCP CallToolResult has .content array
  if (result.content && Array.isArray(result.content)) {
    const texts = result.content
      .filter((c: any) => c.type === 'text' && c.text)
      .map((c: any) => c.text);
    return truncate(texts.join('\n'));
  }
  if (typeof result === 'string') return truncate(result);
  try { return truncate(JSON.stringify(result)); } catch { return ''; }
}

// ────────────────────────────────────────────────────
// Client-side: patch Client.callTool
// ────────────────────────────────────────────────────

export function patchMcpClient(mcpModule: any, debug: boolean): void {
  if (!mcpModule || mcpModule.__trickle_mcp_patched) return;
  mcpModule.__trickle_mcp_patched = true;

  // @modelcontextprotocol/sdk exports Client class
  const ClientClass = mcpModule.Client;
  if (!ClientClass) {
    if (debug) console.log('[trickle/mcp] Client class not found in module');
    return;
  }

  // Patch callTool on the prototype
  const proto = ClientClass.prototype;
  if (proto.callTool && !proto.callTool.__trickle_patched) {
    const origCallTool = proto.callTool;
    proto.callTool = async function patchedCallTool(this: any, ...args: any[]) {
      const params = args[0] || {};
      const toolName = typeof params === 'string' ? params : (params.name || 'unknown');
      const toolArgs = typeof params === 'string' ? args[1] : (params.arguments || params.args);
      const startTime = performance.now();

      try {
        const result = await origCallTool.apply(this, args);
        const durationMs = Math.round((performance.now() - startTime) * 100) / 100;
        writeMcpEvent({
          kind: 'mcp_tool_call', tool: toolName, direction: 'outgoing',
          durationMs, args: sanitizeArgs(toolArgs),
          resultPreview: extractResultPreview(result),
          isError: result?.isError || false,
          errorMessage: result?.isError ? extractResultPreview(result) : undefined,
          timestamp: Date.now(),
        });
        if (debug) console.log(`[trickle/mcp] callTool: ${toolName} (${durationMs}ms)`);
        return result;
      } catch (err: any) {
        const durationMs = Math.round((performance.now() - startTime) * 100) / 100;
        writeMcpEvent({
          kind: 'mcp_tool_call', tool: toolName, direction: 'outgoing',
          durationMs, args: sanitizeArgs(toolArgs),
          resultPreview: '', isError: true,
          errorMessage: truncate(err?.message || String(err), 200),
          timestamp: Date.now(),
        });
        throw err;
      }
    };
    proto.callTool.__trickle_patched = true;
    if (debug) console.log('[trickle/mcp] Patched Client.callTool');
  }

  // Also patch listTools for discovery
  if (proto.listTools && !proto.listTools.__trickle_patched) {
    const origListTools = proto.listTools;
    proto.listTools = async function patchedListTools(this: any, ...args: any[]) {
      const startTime = performance.now();
      const result = await origListTools.apply(this, args);
      const durationMs = Math.round((performance.now() - startTime) * 100) / 100;
      const toolCount = result?.tools?.length || 0;
      writeMcpEvent({
        kind: 'mcp_tool_call', tool: '__list_tools', direction: 'outgoing',
        durationMs, args: null,
        resultPreview: `${toolCount} tools available`,
        isError: false, timestamp: Date.now(),
      });
      return result;
    };
    proto.listTools.__trickle_patched = true;
  }
}

// ────────────────────────────────────────────────────
// Server-side: patch Server to wrap tool handlers
// ────────────────────────────────────────────────────

export function patchMcpServer(mcpModule: any, debug: boolean): void {
  // Server is exported from @modelcontextprotocol/sdk
  const ServerClass = mcpModule.Server || mcpModule.McpServer;
  if (!ServerClass || ServerClass.__trickle_mcp_server_patched) return;
  ServerClass.__trickle_mcp_server_patched = true;

  const proto = ServerClass.prototype;

  // Patch the .tool() registration method to wrap handlers
  if (proto.tool && !proto.tool.__trickle_patched) {
    const origTool = proto.tool;
    proto.tool = function patchedTool(this: any, ...args: any[]) {
      // tool(name, schema, handler) or tool(name, handler)
      const toolName = typeof args[0] === 'string' ? args[0] : 'unknown';
      const lastArg = args[args.length - 1];

      if (typeof lastArg === 'function') {
        const originalHandler = lastArg;
        args[args.length - 1] = async function wrappedHandler(...handlerArgs: any[]) {
          const startTime = performance.now();
          try {
            const result = await originalHandler.apply(this, handlerArgs);
            const durationMs = Math.round((performance.now() - startTime) * 100) / 100;
            writeMcpEvent({
              kind: 'mcp_tool_call', tool: toolName, direction: 'incoming',
              durationMs, args: sanitizeArgs(handlerArgs[0]),
              resultPreview: extractResultPreview(result),
              isError: result?.isError || false,
              timestamp: Date.now(),
            });
            if (debug) console.log(`[trickle/mcp] tool handler: ${toolName} (${durationMs}ms)`);
            return result;
          } catch (err: any) {
            const durationMs = Math.round((performance.now() - startTime) * 100) / 100;
            writeMcpEvent({
              kind: 'mcp_tool_call', tool: toolName, direction: 'incoming',
              durationMs, args: sanitizeArgs(handlerArgs[0]),
              resultPreview: '', isError: true,
              errorMessage: truncate(err?.message || String(err), 200),
              timestamp: Date.now(),
            });
            throw err;
          }
        };
      }

      return origTool.apply(this, args);
    };
    proto.tool.__trickle_patched = true;
    if (debug) console.log('[trickle/mcp] Patched Server.tool');
  }

  // Patch setRequestHandler for lower-level interception
  if (proto.setRequestHandler && !proto.setRequestHandler.__trickle_patched) {
    const origSetHandler = proto.setRequestHandler;
    proto.setRequestHandler = function patchedSetHandler(this: any, schema: any, handler: any) {
      if (typeof handler === 'function') {
        const origHandler = handler;
        const capturedHandler = origHandler;
        handler = async function wrappedHandler(this: any, ...args: any[]) {
          const request = args[0];
          const method = request?.method || schema?.method || 'unknown';
          if (method === 'tools/call') {
            const startTime = performance.now();
            try {
              const result = await capturedHandler.apply(this, args);
              const durationMs = Math.round((performance.now() - startTime) * 100) / 100;
              writeMcpEvent({
                kind: 'mcp_tool_call', tool: request?.params?.name || 'unknown',
                direction: 'incoming', durationMs,
                args: sanitizeArgs(request?.params?.arguments),
                resultPreview: extractResultPreview(result),
                isError: result?.isError || false,
                timestamp: Date.now(),
              });
              return result;
            } catch (err: any) {
              const durationMs = Math.round((performance.now() - startTime) * 100) / 100;
              writeMcpEvent({
                kind: 'mcp_tool_call', tool: request?.params?.name || 'unknown',
                direction: 'incoming', durationMs,
                args: sanitizeArgs(request?.params?.arguments),
                resultPreview: '', isError: true,
                errorMessage: truncate(err?.message || String(err), 200),
                timestamp: Date.now(),
              });
              throw err;
            }
          }
          return capturedHandler.apply(this, args);
        };
      }
      return origSetHandler.call(this, schema, handler);
    };
    proto.setRequestHandler.__trickle_patched = true;
  }
}

// ────────────────────────────────────────────────────
// Initialization
// ────────────────────────────────────────────────────

export function initMcpObserver(): void {
  const dir = process.env.TRICKLE_LOCAL_DIR || path.join(process.cwd(), '.trickle');
  try { fs.mkdirSync(dir, { recursive: true }); } catch {}
  mcpFile = path.join(dir, 'mcp.jsonl');
  try { fs.writeFileSync(mcpFile, ''); } catch {}
  eventCount = 0;
}

File: packages/client-python/src/trickle/mcp_observer.py
"""MCP tool call observer — auto-instruments MCP client and server SDKs.

Captures tool invocations, arguments, responses, latency, and errors.
Writes to .trickle/mcp.jsonl.
"""

from __future__ import annotations

import json
import os
import time
from typing import Any

_debug = False
_mcp_file: str | None = None
_event_count = 0
_MAX_EVENTS = 1000
_TRUNCATE_LEN = 500


def _get_mcp_file() -> str:
    global _mcp_file
    if _mcp_file:
        return _mcp_file
    local_dir = os.environ.get("TRICKLE_LOCAL_DIR") or os.path.join(os.getcwd(), ".trickle")
    os.makedirs(local_dir, exist_ok=True)
    _mcp_file = os.path.join(local_dir, "mcp.jsonl")
    return _mcp_file


def _write_event(event: dict[str, Any]) -> None:
    global _event_count
    if _event_count >= _MAX_EVENTS:
        return
    _event_count += 1
    try:
        with open(_get_mcp_file(), "a") as f:
            f.write(json.dumps(event) + "\n")
    except Exception:
        pass


def _truncate(s: str, length: int = _TRUNCATE_LEN) -> str:
    if not s:
        return ""
    return s[:length] + "..." if len(s) > length else s


def _sanitize_args(args: Any) -> Any:
    if args is None:
        return None
    try:
        s = json.dumps(args)
        return json.loads(s[:1000]) if len(s) > 1000 else args
    except Exception:
        return str(args)[:200]


def _extract_result_preview(result: Any) -> str:
    if not result:
        return ""
    content = getattr(result, "content", None)
    if content and isinstance(content, list):
        texts = []
        for c in content:
            if getattr(c, "type", None) == "text" and getattr(c, "text", None):
                texts.append(c.text)
        return _truncate("\n".join(texts))
    if isinstance(result, str):
        return _truncate(result)
    try:
        return _truncate(json.dumps(result))
    except Exception:
        return ""


# ────────────────────────────────────────────────────
# Client-side: patch ClientSession.call_tool
# ────────────────────────────────────────────────────


def patch_mcp_client(mcp_module: Any) -> None:
    """Patch the MCP Python SDK client to capture tool calls."""
    if getattr(mcp_module, "_trickle_mcp_patched", False):
        return
    mcp_module._trickle_mcp_patched = True

    # mcp.ClientSession has call_tool method
    ClientSession = None
    try:
        from mcp import ClientSession as CS
        ClientSession = CS
    except ImportError:
        try:
            ClientSession = getattr(mcp_module, "ClientSession", None)
        except Exception:
            pass

    if ClientSession is None:
        return

    if hasattr(ClientSession, "call_tool") and not getattr(ClientSession.call_tool, "_trickle_patched", False):
        _orig_call_tool = ClientSession.call_tool

        async def _patched_call_tool(self: Any, name: str, arguments: Any = None, **kwargs: Any) -> Any:
            start = time.perf_counter()
            error_msg = None
            result = None
            try:
                result = await _orig_call_tool(self, name, arguments, **kwargs)
                return result
            except Exception as e:
                error_msg = str(e)[:200]
                raise
            finally:
                duration_ms = round((time.perf_counter() - start) * 1000, 2)
                is_error = bool(error_msg) or (result is not None and getattr(result, "is_error", False))
                _write_event({
                    "kind": "mcp_tool_call",
                    "tool": name,
                    "direction": "outgoing",
                    "durationMs": duration_ms,
                    "args": _sanitize_args(arguments),
                    "resultPreview": _extract_result_preview(result) if not error_msg else "",
                    "isError": is_error,
                    "errorMessage": error_msg or (_extract_result_preview(result) if is_error else None),
                    "timestamp": int(time.time() * 1000),
                })
                if _debug:
                    print(f"[trickle/mcp] call_tool: {name} ({duration_ms}ms)")

        _patched_call_tool._trickle_patched = True  # type: ignore
        ClientSession.call_tool = _patched_call_tool

    # Also patch list_tools
    if hasattr(ClientSession, "list_tools") and not getattr(ClientSession.list_tools, "_trickle_patched", False):
        _orig_list_tools = ClientSession.list_tools

        async def _patched_list_tools(self: Any, **kwargs: Any) -> Any:
            start = time.perf_counter()
            result = await _orig_list_tools(self, **kwargs)
            duration_ms = round((time.perf_counter() - start) * 1000, 2)
            tool_count = len(getattr(result, "tools", []) or [])
            _write_event({
                "kind": "mcp_tool_call",
                "tool": "__list_tools",
                "direction": "outgoing",
                "durationMs": duration_ms,
                "args": None,
                "resultPreview": f"{tool_count} tools available",
                "isError": False,
                "timestamp": int(time.time() * 1000),
            })
            return result

        _patched_list_tools._trickle_patched = True  # type: ignore
        ClientSession.list_tools = _patched_list_tools

    if _debug:
        print("[trickle/mcp] Patched MCP ClientSession")


# ────────────────────────────────────────────────────
# Server-side: patch FastMCP tool decorator
# ────────────────────────────────────────────────────


def patch_mcp_server(mcp_module: Any) -> None:
    """Patch the MCP Python SDK server to capture incoming tool calls."""
    if getattr(mcp_module, "_trickle_mcp_server_patched", False):
        return
    mcp_module._trickle_mcp_server_patched = True

    # FastMCP has a .tool() decorator
    FastMCP = None
    try:
        from mcp.server.fastmcp import FastMCP as FM
        FastMCP = FM
    except ImportError:
        pass

    if FastMCP is None:
        return

    if hasattr(FastMCP, "tool") and not getattr(FastMCP.tool, "_trickle_patched", False):
        _orig_tool = FastMCP.tool

        def _patched_tool(self: Any, *args: Any, **kwargs: Any) -> Any:
            decorator = _orig_tool(self, *args, **kwargs)

            def _wrapping_decorator(fn: Any) -> Any:
                registered = decorator(fn)
                tool_name = getattr(fn, "__name__", "unknown")

                # Wrap the registered function
                import functools
                import asyncio

                if asyncio.iscoroutinefunction(fn):
                    @functools.wraps(fn)
                    async def _wrapped(*a: Any, **kw: Any) -> Any:
                        start = time.perf_counter()
                        error_msg = None
                        result = None
                        try:
                            result = await fn(*a, **kw)
                            return result
                        except Exception as e:
                            error_msg = str(e)[:200]
                            raise
                        finally:
                            duration_ms = round((time.perf_counter() - start) * 1000, 2)
                            _write_event({
                                "kind": "mcp_tool_call",
                                "tool": tool_name,
                                "direction": "incoming",
                                "durationMs": duration_ms,
                                "args": _sanitize_args(kw or (a[0] if a else None)),
                                "resultPreview": _truncate(str(result)[:500]) if result and not error_msg else "",
                                "isError": bool(error_msg),
                                "errorMessage": error_msg,
                                "timestamp": int(time.time() * 1000),
                            })
                    return _wrapped
                else:
                    @functools.wraps(fn)
                    def _wrapped_sync(*a: Any, **kw: Any) -> Any:
                        start = time.perf_counter()
                        error_msg = None
                        result = None
                        try:
                            result = fn(*a, **kw)
                            return result
                        except Exception as e:
                            error_msg = str(e)[:200]
                            raise
                        finally:
                            duration_ms = round((time.perf_counter() - start) * 1000, 2)
                            _write_event({
                                "kind": "mcp_tool_call",
                                "tool": tool_name,
                                "direction": "incoming",
                                "durationMs": duration_ms,
                                "args": _sanitize_args(kw or (a[0] if a else None)),
                                "resultPreview": _truncate(str(result)[:500]) if result and not error_msg else "",
                                "isError": bool(error_msg),
                                "errorMessage": error_msg,
                                "timestamp": int(time.time() * 1000),
                            })
                    return _wrapped_sync

            # If original .tool() is used as a bare decorator (no args), handle that
            if args and callable(args[0]):
                return _wrapping_decorator(args[0])
            return _wrapping_decorator

        _patched_tool._trickle_patched = True  # type: ignore
        FastMCP.tool = _patched_tool

    if _debug:
        print("[trickle/mcp] Patched FastMCP server")


# ────────────────────────────────────────────────────
# Installation
# ────────────────────────────────────────────────────


def patch_mcp(debug: bool = False) -> None:
    """Install MCP observer hooks."""
    global _debug
    _debug = debug

    import sys

    # Clear previous data
    try:
        f = _get_mcp_file()
        with open(f, "w") as fp:
            fp.truncate(0)
    except Exception:
        pass

    # Patch already-imported modules
    if "mcp" in sys.modules:
        try:
            patch_mcp_client(sys.modules["mcp"])
        except Exception:
            pass
    if "mcp.server.fastmcp" in sys.modules:
        try:
            patch_mcp_server(sys.modules["mcp.server.fastmcp"])
        except Exception:
            pass

    # Register with the consolidated import hook
    try:
        from trickle.db_observer import register_import_patches
        register_import_patches({
            "mcp": patch_mcp_client,
            "mcp.server.fastmcp": patch_mcp_server,
        })
    except Exception:
        pass
